detox/src/client/actions/SyncStatusFormatter.js

```diff
@@ -0,0 +1,186 @@
+const Ajv = require('ajv');
+
+const DetoxInternalError = require('../../errors/DetoxInternalError');
+const statusSchema = require('../actions/SyncStatusSchema.json');
+
+
+const ajv = new Ajv();
+const isValidJSONStatus = ajv.compile(statusSchema);
+
+function formatJSONStatus(jsonStatus) {
+  if (!isValidJSONStatus(jsonStatus)) {
+    const errorMessages = isValidJSONStatus.errors.map(
+      error => `• ${error.message} in path "${error.schemaPath}" with params: ${JSON.stringify(error.params)}`
+    );
+    throw new DetoxInternalError(`Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), ` +
+      `given status: ${JSON.stringify(jsonStatus)}.\nWith reasons:\n${errorMessages.join('\n')}\n`);
+  }
+
+  if (isAppIdle(jsonStatus)) {
+    return 'The app seems to be idle';
+  }
+
+  const resourcesDescriptions = resourcesDescriptionsFromJSON(jsonStatus.busy_resources);
+  return `The app is busy with the following tasks:\n${resourcesDescriptions.join('\n')}`;
+}
+
+function isAppIdle(jsonStatus) {
+  return jsonStatus.app_status === 'idle';
+}
+
+function resourcesDescriptionsFromJSON(jsonDescriptions) {
+  let descriptions = [];
+  for (const jsonDescription of jsonDescriptions) {
+    const description = resourceDescriptionFromJSON(jsonDescription);
+    descriptions.push(description);
+  }
+
+  return descriptions;
+}
+
+const resourceFormatters = {
+  delayed_perform_selector: delayedPerformSelectorResourceDescriptionFromProperties,
```

Comment:
I think that maybe these should be extracted onto separate files (without necessarily breaking the unit test into several ones). I think it would help with:
- Shortening the various function names (for example, `delayedPerformSelectorResourceDescriptionFromProperties` => `propertiesFormatter.delayedPerformSelector`).
- Putting things like `uiResourceCounterNameMapping` more intimately alongside their concern.

Reply:
Sounds great :+1:

Reply:
done, extracted to separated formatters.